.github/workflows/docker-publish.yml

```diff
@@ -0,0 +1,115 @@
+# This is a basic workflow to help you get started with Actions
+
+name: Docker Build and Publish
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the main branch
+on:
+  schedule:
+    - cron: '0 0 * * *' # Nightly build
+  pull_request:
+    branches: [ master ] 
+  push:
+    branches: [ master ]
+
+  workflow_dispatch: 
+    
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  # This workflow contains a single job called "build"
+  build_and_push:
+    name: Build, tag, and push images
+    env:
+      default_python: "3.9"
+
+    strategy:
+      fail-fast: true
+      matrix:
+        python_images: [3.6,3.7,3.8,3.9]
+    # The type of runner that the job will run on
+    runs-on: ubuntu-latest
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Check that appropriate secrets are specified
+      # https://github.com/actions/runner/issues/520
+      -
+        name: Check Secrets
+        run: |
+          if [ "x${{ secrets.DOCKER_SOPEL_IMAGE_NAME }}" == "x" ] || [ "x${{ secrets.DOCKER_HUB_USERNAME }}" == "x" ] || [ "x${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}" == "x" ]; then 
+            echo "Required secrets not set!"
+            exit 1
+          fi
+
```

Comment:
```suggestion
```

The goal of this step was to prevent a pile up of failed jobs. Ultimately, the workflow should not even run if the secrets are missing, not prevent subsequent steps by failling, as is happening here. `exit 0` would have been acceptable, but not optimal.

Since we're in here making changes, lets just go with https://github.community/t/how-can-i-test-if-secrets-are-available-in-an-action/17911/11.

That way, we can avoid extra steps _AND_ having to use sh scripting in one go.

So basically, the final requested changes are, 
1) set workflow _or_ job level ENVs (your choice) 
  ```yaml
  …
  env:
    DOCKER_SOPEL_IMAGE_NAME: ${{ secrets.DOCKER_SOPEL_IMAGE_NAME }}
    DOCKER_HUB_USERNAME: …
    …
  …
  ```
_AND_

2) include a job level 
   ```yaml
   if: ${{ (env.DOCKER_SOPEL_IMAGE_NAME != '') && (env.DOCKER_HUB_USERNAME != '') && (env.DOCKER_HUB_ACCESS_TOKEN != '') }}
     ```
.

Reply:
The env at the workflow or job level don't seem to work. So unless you want an `if` at every step I think following the link (https://github.community/t/how-can-i-test-if-secrets-are-available-in-an-action/17911/11) and making it a pre-flight stage

Reply:
for posterity:
```
14:37:02 <+HumorBaby> dagnabbit adamus1red, this GHA workflow condition is a real PITA
14:38:06 <+HumorBaby> honestly, I think it should go back to my original suggestion of if: ${{ github.repository_owner == 'sopel-irc' }} 
14:39:36 <+HumorBaby> anyone who forks is likely going to have their own changes (if not, they can use the base image), and in that case, they can just perma-float (append) the commit to update the GHA workflow file to their branch/tree as I desribed
14:40:55 <+HumorBaby> since there is no way to cleanly condition the 'start' of a job without at least 1 failure (when secrets are missing), forks are still going to be flooded with nightly failed workflows
14:41:45 <+HumorBaby> at that point, they will modify their actions yaml anyway, so might as well have included the owner check anyway
14:45:18 <adamus1red> HumorBaby: forks need to manually enable actions
14:45:34 <adamus1red> so it's not like all 4 forks are going to get spammed instantly
14:45:58 <+HumorBaby> lol 4 forks
14:47:54 <+HumorBaby> hmm ok I'll take your word for it for now
14:49:18 <+HumorBaby> so then lets just scrap that check all together. Everything else looked good so looks like we'll merge after this final change
```